mbf_abstract_nav/src/controller_action.cpp

```diff
@@ -51,59 +51,38 @@ ControllerAction::ControllerAction(
 }
 
 void ControllerAction::start(
-    GoalHandle goal_handle,
+    GoalHandle &goal_handle,
     typename AbstractControllerExecution::Ptr execution_ptr
 )
 {
-  bool new_plan = false;
-  boost::lock_guard<boost::mutex> guard(map_mtx_);
-  typename SlotGoalIdMap::left_const_iterator slot
-      = concurrency_slots_.left.find(goal_handle.getGoal()->concurrency_slot);
-  if(slot != concurrency_slots_.left.end())
+  uint8_t slot = goal_handle.getGoal()->concurrency_slot;
+
+  std::map<uint8_t, ConcurrencySlot>::iterator slot_it = concurrency_slots_.find(slot);
+  if(slot_it != concurrency_slots_.end())
   {
+    boost::lock_guard<boost::mutex> guard(slots_mtx_);
 
-    typename std::map<const std::string, const typename AbstractControllerExecution::Ptr>::const_iterator elem
-        = executions_.find(slot->second);
-    if(elem != executions_.end())
+    if(slot_it->second.execution->getName() == goal_handle.getGoal()->controller ||
+       goal_handle.getGoal()->controller.empty())
     {
-      if(elem->second->getName() == goal_handle.getGoal()->controller || goal_handle.getGoal()->controller.empty())
-      {
-        execution_ptr = elem->second;
-        execution_ptr->setNewPlan(goal_handle.getGoal()->path.poses);
-        new_plan = true;
-      }else{
-        elem->second->cancel();
-        concurrency_slots_.left.erase(slot->first);
-      }
+      // Goal requests to run the same controller on the same concurrency slot:
+      // we update the goal handle and pass the new plan to the execution without stopping it
+      execution_ptr = slot_it->second.execution;
+      execution_ptr->setNewPlan(goal_handle.getGoal()->path.poses);
+      concurrency_slots_[slot].goal_handle = goal_handle;
+      return;
```

Comment:
Use else instead -> more readable

Reply:
mmm...  yes, but would be more complex structure, as I want to prevent the execution of the _AbstractAction::start(goal_handle, execution_ptr)_ if any of the ifs is false (so either 2 elses with the call repeated, or a flag, both uglier than the return, imho)

Reply:
Maybe you need to define a ref in the struct and hand over the ref via a constructor.